internal/services/pam/pam_test.go

```diff
@@ -0,0 +1,595 @@
+package pam_test
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"github.com/ubuntu/authd"
+	"github.com/ubuntu/authd/internal/brokers"
+	"github.com/ubuntu/authd/internal/services/pam"
+	"github.com/ubuntu/authd/internal/testutils"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+var (
+	testClient        authd.PAMClient
+	brokerGeneratedID string
+)
+
+// Used for TestGetAuthenticationModes and TestSelectAuthenticationMode.
+var (
+	requiredEntries = "required:entry_type,other_entry_type"
+	optionalEntries = "optional:entry_type,other_entry_type"
+	optional        = "optional"
+
+	requiredEntry = &authd.UILayout{
+		Type:    "required-entry",
+		Label:   &optional,
+		Button:  &optional,
+		Wait:    &optional,
+		Entry:   &requiredEntries,
+		Content: &optional,
+	}
+	optionalEntry = &authd.UILayout{
+		Type:  "optional-entry",
+		Entry: &optionalEntries,
+	}
+	emptyType = &authd.UILayout{
+		Type:  "",
+		Entry: &requiredEntries,
+	}
+)
+
+func TestAvailableBrokers(t *testing.T) {
+	t.Parallel()
+
+	abResp, err := testClient.AvailableBrokers(context.Background(), &authd.Empty{})
+	require.NoError(t, err, "AvailableBrokers should not return an error, but did")
+
+	got := abResp.GetBrokersInfos()
+	for _, broker := range got {
+		broker.Id = broker.Name + "_ID"
+	}
+	want := testutils.LoadWithUpdateFromGoldenYAML(t, got)
+	require.Equal(t, want, got, "AvailableBrokers returned unexpected brokers")
+}
+
+func TestGetPreviousBroker(t *testing.T) {
+	t.Parallel()
+
+	username := t.Name()
+
+	// Try to get the broker for the user before assigning it.
+	gotResp, _ := testClient.GetPreviousBroker(context.Background(), &authd.GPBRequest{Username: username})
+	require.Empty(t, gotResp.GetPreviousBroker(), "GetPreviousBroker should return nil when the user has no broker assigned")
+
+	_, err := testClient.SetDefaultBrokerForUser(context.Background(), &authd.SDBFURequest{
+		BrokerId: "local",
+		Username: username,
+	})
+	require.NoError(t, err, "Setup: could not set default broker for user for tests")
+
+	// Assert that the broker assigned to the user is correct.
+	gotResp, _ = testClient.GetPreviousBroker(context.Background(), &authd.GPBRequest{Username: username})
+	require.Equal(t, "local", gotResp.GetPreviousBroker(), "GetPreviousBroker did not return the correct broker")
+
+	// Asset that trying to get the broker from a user with no broker assigned returns empty.
+	gotResp, _ = testClient.GetPreviousBroker(context.Background(), &authd.GPBRequest{Username: username + "no broker"})
```

Comment:
This is then handled in the first call though, is it still necessary?

Reply:
I wouldn't say it's "necessary", but more like a second assertion :joy: I do agree that it can be removed, though.